### PR TITLE
Added support for multiple entities for Go

### PR DESCRIPTION
**Pull request by ygorelik**
Added support for multiple entities for Go

### CODE REVIEW

File: sdk/go/core/docsgen/api/types/ydk_types.rst [hunk 1]
```diff
@@ -268,44 +268,172 @@ These are how YANG types are represented in Go.
 
     An interface type that represents a basic container in YANG
 
-    .. function:: GetEntityData()
+    .. function:: GetEntityData() Entity
```
Comment (ylil93) [inline sdk/go/core/docsgen/api/types/ydk_types.rst]: This function doesn't return Entity; it returns *CommonEntityData. The line should be
`.. function:: GetEntityData() *CommonEntityData`

File: sdk/go/core/docsgen/api/types/ydk_types.rst [hunk 2]
```diff
@@ -268,44 +268,172 @@ These are how YANG types are represented in Go.
...
+    
+        The :ref:`Entity <types-entity>` interface function
```
Comment (ylil93) [inline sdk/go/core/docsgen/api/types/ydk_types.rst]: This is vague. Instead say something like
Implements :ref:\`Entity <types-entity>\` interface

File: sdk/go/core/docsgen/api/types/ydk_types.rst [hunk 3]
```diff
@@ -268,44 +268,172 @@ These are how YANG types are represented in Go.
...
+
+.. go:struct:: EntityCollection
+
+    Type `EntityCollection` along with its methods implements ordered map collection of entities. The string value of entity `SegmentPath` serves as a map key for the entity. Ordered means, the collection retains order of entities, in which they were added to collection. 
+    
+    The `EntityCollection` type has two aliases - `Config` and `Filter`.
+
+    .. function:: GetEntityData() *CommonEntityData
+
+        Implements Entity interface.
```
Comment (ylil93) [inline sdk/go/core/docsgen/api/types/ydk_types.rst]: I think it's better to include the link to to the Entity interface:

Implements :ref:\`Entity <types-entity>\` interface

File: sdk/go/core/tests/service_crud_test.go [hunk 4]
```diff
@@ -3,7 +3,8 @@ package test
 import (
 	"fmt"
 	ysanity "github.com/CiscoDevNet/ydk-go/ydk/models/ydktest/sanity"
-	// "github.com/CiscoDevNet/ydk-go/ydk/models/ietf/interfaces"
```
Comment (ylil93) [inline sdk/go/core/tests/service_crud_test.go]: If this line isn't needed, please delete it.

File: sdk/go/core/ydk/path/path.go [hunk 5]
```diff
@@ -253,25 +262,57 @@ func getTopEntityFromFilter(filter types.Entity) types.Entity {
...
-		fmt.Sprintf("Reading top entity: '%s'", segmentPath))
-
 	cchildren := C.DataNodeGetChildren(readDataNode.Private.(C.DataNode))
 
 	if cchildren.count == C.int(0) {
-		return topEntity
+		return filter;
 	}
-
+
```
Comment (ylil93) [inline sdk/go/core/ydk/path/path.go]: Extra indentation should be deleted

File: sdk/go/core/ydk/types/types.go [hunk 6]
```diff
@@ -103,14 +103,22 @@ type CommonEntityData struct {
 	BundleYangModelsLocation	string
 
 	// dynamic data (internals)
-	Parent 						Entity
+	Parent 					Entity
+}
+
+func (e *CommonEntityData) GetPath() string {
```
Comment (ylil93) [inline sdk/go/core/ydk/types/types.go]: Is this method necessary? For the purpose of extracting the segment path from an Entity struct, this method exists
https://github.com/ygorelik/ydk-gen/blob/715bed51b22bf3cb9e1e8d69aee23f9ee2d9544e/sdk/go/core/ydk/types/types.go#L316

File: sdk/cpp/core/src/path/netconf_session.cpp [hunk 7]
```diff
@@ -489,8 +489,9 @@ static string get_filter_payload(path::Rpc & ydk_rpc)
 {
     auto entity = ydk_rpc.get_input_node().find("filter");
     if(entity.empty()){
-        YLOG_ERROR("Failed to get entity node.");
-        throw(YInvalidArgumentError{"Failed to get entity node"});
+        YLOG_WARN("Failed to get filter node for RPC; it then will be skipped.");
```
Comment (abhikeshav) [inline sdk/cpp/core/src/path/netconf_session.cpp]: Why is this changed to warning?

File: sdk/cpp/core/src/ydk.cpp [hunk 8]
```diff
@@ -117,21 +117,23 @@ static ydk::EncodingFormat get_real_encoding(EncodingFormat encoding)
 {
     switch(encoding)
     {
-        case XML:
-            return ydk::EncodingFormat::XML;
         case JSON:
             return ydk::EncodingFormat::JSON;
+        case XML:
+        default:
```
Comment (abhikeshav) [inline sdk/cpp/core/src/ydk.cpp]: `default` `case`s should be avoided. Suggest removing this or adding cases for all the members of the enum

File: sdk/cpp/core/src/ydk.cpp [hunk 9]
```diff
@@ -117,21 +117,23 @@ static ydk::EncodingFormat get_real_encoding(EncodingFormat encoding)
...
 static ydk::Protocol get_real_protocol(Protocol protocol)
 {
     switch(protocol)
     {
-        case Netconf:
-            return ydk::Protocol::netconf;
         case Restconf:
             return ydk::Protocol::restconf;
+        case Netconf:
+        default:
```
Comment (abhikeshav) [inline sdk/cpp/core/src/ydk.cpp]: Same as above. Better to avoid default switch cases

File: sdk/go/README.md [hunk 10]
```diff
@@ -64,6 +64,16 @@ $ sudo yum install https://devhub.cisco.com/artifactory/rpm-ydk/0.7.1/libydk-0.7
 
 ```
 
+Golang
```
Comment (abhikeshav) [inline sdk/go/README.md]: Can this be made to a heading? Something like

```
### Golang version
```

File: sdk/go/core/docsgen/conf.py [hunk 11]
```diff
@@ -36,6 +32,8 @@
 
 sys.path.append(os.path.abspath(ydk_root+'/sphinxexts'))
 extensions = [ 'godomain' ]
+print(extensions, sys.path)
```
Comment (abhikeshav) [inline sdk/go/core/docsgen/conf.py]: Can this print statement be removed?

File: sdk/go/core/docsgen/api/types/ydk_types.rst (hunk 2 above)
Comment (ygorelik) [reply sdk/go/core/docsgen/api/types/ydk_types.rst]: Fixed.

File: sdk/go/core/docsgen/api/types/ydk_types.rst (hunk 3 above)
Comment (ygorelik) [reply sdk/go/core/docsgen/api/types/ydk_types.rst]: Fixed.

File: sdk/go/core/tests/service_crud_test.go (hunk 4 above)
Comment (ygorelik) [reply sdk/go/core/tests/service_crud_test.go]: Done

File: sdk/go/core/ydk/path/path.go (hunk 5 above)
Comment (ygorelik) [reply sdk/go/core/ydk/path/path.go]: Will check-in with next commit due to code overlapping with project in work.

File: sdk/go/core/ydk/types/types.go (hunk 6 above)
Comment (ygorelik) [reply sdk/go/core/ydk/types/types.go]: Deleted. Will check-in with next commit due to code overlapping with project in work.

File: sdk/cpp/core/src/path/netconf_session.cpp (hunk 7 above)
Comment (ygorelik) [reply sdk/cpp/core/src/path/netconf_session.cpp]: It is not an error when filter is not present. In this case the whole device configuration will be fetched.

File: sdk/cpp/core/src/ydk.cpp (hunk 8 above)
Comment (ygorelik) [reply sdk/cpp/core/src/ydk.cpp]: The compiler logged warning message pointing that 'default' case is missing. I was taught always include 'default' case. If not included the 'switch' must be re-factored to if-else statements.

File: sdk/cpp/core/src/ydk.cpp (hunk 9 above)
Comment (ygorelik) [reply sdk/cpp/core/src/ydk.cpp]: The compiler logged warning message pointing that 'default' case is missing. I was taught always include 'default' case. If not included the 'switch' must be re-factored to if-else statements.

File: sdk/cpp/tests/test_netconf_service.cpp [hunk 12]
```diff
@@ -328,30 +328,3 @@ TEST_CASE("ietf_get_rpc")
         cout << "Exception while executing RPC: " << ex.what() << endl;
     }
 }
-
-TEST_CASE("get_config_openconfig_interfaces_and_bgp")
```
[the remark this answers is not in the document]
Comment (ylil93) [reply sdk/cpp/tests/test_netconf_service.cpp]: I see. It's probably better to comment it out with that explanation rather than deleting everything.

See this example:
https://github.com/ygorelik/ydk-gen/blob/715bed51b22bf3cb9e1e8d69aee23f9ee2d9544e/sdk/cpp/tests/test_netconf_service.cpp#L40

File: sdk/go/core/docsgen/api/types/ydk_types.rst (hunk 2 above)
Comment (ylil93) [reply sdk/go/core/docsgen/api/types/ydk_types.rst]: I did not see the change in your latest push - bumping this comment so it doesn't get buried

File: sdk/go/core/docsgen/api/types/ydk_types.rst (hunk 3 above)
Comment (ylil93) [reply sdk/go/core/docsgen/api/types/ydk_types.rst]: I did not see the change in your latest push - bumping this comment so it doesn't get buried

Comment (ygorelik) [reply sdk/go/core/docsgen/api/types/ydk_types.rst]: It is in my fork: https://github.com/ygorelik/ydk-gen/commit/d227bd9997518478a9f1055198c7f38cc07b5fc6
